### PR TITLE
8111/feature/rename and move build marc

### LINKED ISSUE #8111: Rename `build_marc` and move it to a new home
`build_marc` in [`catalog/merge/merge_marc.py`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/catalog/merge/merge_marc.py) is poorly named. 

### Proposal & Constraints

@hornc has [suggested](https://github.com/internetarchive/openlibrary/pull/7940#discussion_r1228968439) renaming this to `expand_record()` and moving it to a new home. How about [`catalog/utils/__init__.py`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/catalog/utils/__init__.py)

There is perhaps a dangerous thread to pull at here, insofar as [`catalog/merge/merge_marc.py`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/catalog/merge/merge_marc.py)
and [`catalog/merge/merge.py`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/catalog/merge/merge.py) have a number of overlapping functions. Some of those that overlap have been decorated as `@deprecated` in `merge.py`, but not all.

To avoid getting bogged down too far, let's limit this issue to dealing with `expand_record()`, nee `build_marc()`, and its direct dependencies (i.e. `build_titles()`).

Perhaps both can go from `merge_marc.py` to [`catalog/utils/__init__.py`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/catalog/utils/__init__.py), and then the `build_titles()` from `merge.py` can be decorated with `@deprecated` and it can point to the new location of `build_titles()`.

### Stakeholders

@hornc, @cclauss

### PR DESCRIPTION
**Pull request by scottbarnes**
Closes #8111

This renames `build_marc()` to `expand_record()`, as mentioned [here](https://github.com/internetarchive/openlibrary/pull/7940#discussion_r1228968439), and moves it to [`catalog/utils/__init__.py`](https://github.com/internetarchive/openlibrary/blob/master/openlibrary/catalog/utils/__init__.py).

### Technical

I am not sure if more should be moved. I moved [`build_titles()`](https://github.com/internetarchive/openlibrary/blob/a9d297158999f2d66550c911534349e392abb96c/openlibrary/catalog/merge/merge_marc.py#L17-L53), upon which `expand_record()` depends, around repeatedly, deleted it from [`merge.py`](https://github.com/internetarchive/openlibrary/blob/a9d297158999f2d66550c911534349e392abb96c/openlibrary/catalog/merge/merge.py#L18-L47) repeatedly, moved it back. Moved it again. Then I move other things, and then moved them back. And then moved everything again, etc.. In the end, I think I was just making a mess and kept this small.

If I picked a bad new home for `expand_record()`, just let me know and I'll try another. :)

### Testing

`git grep build_marc` to verify all the `build_marc()` references have been located and updated.

### Stakeholders

@hornc
@cclauss

### CODE REVIEW

File: openlibrary/catalog/add_book/match.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import web
 from deprecated import deprecated
+from openlibrary.catalog.utils import expand_record
 from openlibrary.catalog.merge.merge_marc import (
-    build_marc,
     editions_match as threshold_match,
```
Comment (cclauss) [inline openlibrary/catalog/add_book/match.py:5]: Remove this comma so psf/black can put this import on a single line. ;-)

```suggestion
    editions_match as threshold_match
```

Comment (scottbarnes) [reply openlibrary/catalog/add_book/match.py]: Ahhh fail. Thanks for catching that!

File: openlibrary/tests/catalog/test_utils.py [hunk 2]
```diff
@@ -209,3 +210,18 @@ def test_mk_norm(title, expected):
...
+    edition = {
+        'title': 'A test title (parens)',
+        'full_title': 'A test full title : subtitle (parens).',  # required, and set by add_book.load()
+        'source_records': ['ia:test-source'],
+    }
+    result = expand_record(edition)
+    assert isinstance(result['titles'], list)
+    assert result['isbn'] == []
+    assert result['normalized_title'] == 'a test full title subtitle (parens)'
+    assert result['short_title'] == 'a test full title subtitl'
```
Comment (cclauss) [inline openlibrary/tests/catalog/test_utils.py:227]: Not yet tested...
```suggestion
    expanded_record = expand_record(edition)
    assert isinstance(expanded_record['titles'], list)
    assert expanded_record['isbn'] == []
    assert expanded_record['normalized_title'] == 'a test full title subtitle (parens)'
    assert expanded_record['short_title'] == 'a test full title subtitl'
    assert 'publish_country' not in expanded_record
    for publish_country in {'   ', '|||'}:
        edition['publish_country'] = publish_country
        assert 'publish_country' not in expand_record(edition)
    for publish_country in {'USA', 'usa'}:
        edition['publish_country'] = publish_country
        assert expand_record(edition)['publish_country' == publish_country    
```

Comment (cclauss) [inline openlibrary/tests/catalog/test_utils.py:227]: Also, please add `def test_expand_record_isbns()`...

```suggestion
valid_edition = {
    'title': 'A test title (parens)',
    'full_title': 'A test full title : subtitle (parens).',  # required, and set by add_book.load()
    'source_records': ['ia:test-source'],
}
def test_expand_record():
    # used in openlibrary.catalog.add_book.load()
    # when trying to find an existing edition match
    edition = valid_edition.copy()
    expanded_record = expand_record(edition)
    assert isinstance(expanded_record['titles'], list)
    assert expanded_record['titles'] == [
        'A test full title : subtitle (parens).',
        'a test full title subtitle (parens)',
        'test full title : subtitle (parens).',
        'test full title subtitle (parens)'
    ]
    assert expanded_record['isbn'] == []
    assert expanded_record['normalized_title'] == 'a test full title subtitle (parens)'
    assert expanded_record['short_title'] == 'a test full title subtitl'

def test_expand_record_publish_country():
    # used in openlibrary.catalog.add_book.load()
    # when trying to find an existing edition match
    edition = valid_edition.copy()
    expanded_record = expand_record(edition)
    assert 'publish_country' not in expanded_record
    for publish_country in ('   ', '|||'):
        edition['publish_country'] = publish_country
        assert 'publish_country' not in expand_record(edition)
    for publish_country in ('USA', 'usa'):
        edition['publish_country'] = publish_country
        assert expand_record(edition)['publish_country'] == publish_country

def test_expand_record_transfer_fields():
    edition = valid_edition.copy()
    expanded_record = expand_record(edition)
    transfer_fields = (
        'lccn',
        'publishers',
        'publish_date',
        'number_of_pages',
        'authors',
        'contribs',
    )
    for field in transfer_fields:
        assert field not in expanded_record
    for field in transfer_fields:
        edition[field] = field
    expanded_record = expand_record(edition)
    for field in transfer_fields:
        assert field in expanded_record
```

Comment (scottbarnes) [reply openlibrary/tests/catalog/test_utils.py]: I added these tests and a test for `isbn`.

I confess, much to my shame, I didn't really look at the content of `expand_record`/`build_marc` or that of the tests as I was moving them. Having made a test for the `isbn` field, I wonder if that field is even used anywhere. I will look into this a bit more and report back.

**Review by cclauss — APPROVED** [review]
LGTM

File: openlibrary/tests/catalog/test_utils.py (hunk 2 above)
Comment (scottbarnes) [reply openlibrary/tests/catalog/test_utils.py]: I ran out of steam trying to determine if the `isbn` field is used, given it is not currently causing any problems. My inclination is to go with the added test for it and drop the matter on the theory that if it isn't broken, perhaps I ought not try to fix it.

**Review by cclauss — APPROVED** [review]

**Pull request merged**